From 971533b1790fe38677b9140a210347f7ef6c044a Mon Sep 17 00:00:00 2001
From: Tong Wu <tong1.wu@intel.com>
Date: Mon, 4 Jul 2022 16:09:54 +0800
Subject: [PATCH 2/5] avfilter/vf_hwmap: get the AVHWDeviceType from outlink
 format

When both derive_device_type and device context are not
presented during hw->hw map, the hwmap filter should be
able to retrieve AVHWDeviceType from outlink->format and create
corresponding hwdevice context.

Signed-off-by: Tong Wu <tong1.wu@intel.com>
---
 libavfilter/vf_hwmap.c | 35 ++++++++++++++++++++++-------------
 1 file changed, 22 insertions(+), 13 deletions(-)

diff --git a/libavfilter/vf_hwmap.c b/libavfilter/vf_hwmap.c
index e246b22603..e5b40233b2 100644
--- a/libavfilter/vf_hwmap.c
+++ b/libavfilter/vf_hwmap.c
@@ -70,16 +70,31 @@ static int hwmap_config_output(AVFilterLink *outlink)
     device_is_derived = 0;
 
     if (inlink->hw_frames_ctx) {
+        enum AVHWDeviceType type;
         hwfc = (AVHWFramesContext*)inlink->hw_frames_ctx->data;
 
-        if (ctx->derive_device_type) {
-            enum AVHWDeviceType type;
+        desc = av_pix_fmt_desc_get(outlink->format);
+        if (!desc) {
+            err = AVERROR(EINVAL);
+            goto fail;
+        }
 
-            type = av_hwdevice_find_type_by_name(ctx->derive_device_type);
-            if (type == AV_HWDEVICE_TYPE_NONE) {
-                av_log(avctx, AV_LOG_ERROR, "Invalid device type.\n");
-                err = AVERROR(EINVAL);
-                goto fail;
+        if (ctx->derive_device_type || (!device && (desc->flags & AV_PIX_FMT_FLAG_HWACCEL))) {
+            if (ctx->derive_device_type) {
+                type = av_hwdevice_find_type_by_name(ctx->derive_device_type);
+                if (type == AV_HWDEVICE_TYPE_NONE) {
+                    av_log(avctx, AV_LOG_ERROR, "Invalid device type.\n");
+                    err = AVERROR(EINVAL);
+                    goto fail;
+                }
+            } else {
+                type = av_hwdevice_get_type_by_pix_fmt(outlink->format);
+                if (type == AV_HWDEVICE_TYPE_NONE) {
+                    av_log(avctx, AV_LOG_ERROR, "Could not get device type from "
+                        "format %s.\n", av_get_pix_fmt_name(outlink->format));
+                    err = AVERROR(EINVAL);
+                    goto fail;
+                }
             }
 
             err = av_hwdevice_ctx_create_derived(&device, type,
@@ -92,12 +107,6 @@ static int hwmap_config_output(AVFilterLink *outlink)
             device_is_derived = 1;
         }
 
-        desc = av_pix_fmt_desc_get(outlink->format);
-        if (!desc) {
-            err = AVERROR(EINVAL);
-            goto fail;
-        }
-
         if (inlink->format == hwfc->format &&
             (desc->flags & AV_PIX_FMT_FLAG_HWACCEL) &&
             !ctx->reverse) {
-- 
2.41.0.windows.1

